lib/EarlyFunctionAnalysis/EarlyFunctionAnalysis.cpp

```diff
@@ -843,6 +875,105 @@ void FunctionEntrypointAnalyzer::interproceduralPropagation() {
+                                                .ArgumentsRegisters))
+      Map[GPRArg].IsUsedForPassingArguments = *MaybeState;
+
+  for (auto &GPRRetVal : GPRRetVals)
+    if (auto MaybeState = TryGetRegisterState(GPRRetVal,
+                                              Summary.ABIResults
+                                                .FinalReturnValuesRegisters))
+      Map[GPRRetVal].IsUsedForReturningValues = *MaybeState;
+
+  for (auto &VRArg : VRArguments)
+    if (auto MaybeState = TryGetRegisterState(VRArg,
+                                              Summary.ABIResults
+                                                .ArgumentsRegisters))
+      Map[VRArg].IsUsedForPassingArguments = *MaybeState;
+
+  for (auto &VRRetVal : VRRetVals)
+    if (auto MaybeState = TryGetRegisterState(VRRetVal,
+                                              Summary.ABIResults
+                                                .FinalReturnValuesRegisters))
+      Map[VRRetVal].IsUsedForReturningValues = *MaybeState;
+}
+
+void FunctionEntrypointAnalyzer::applyABIDeductions() {
+  using State = abi::RegisterState::Values;
+
+  if (!ABIEnforcement.getNumOccurrences() || ABIEnforcement == NoABIEnforcement)
+    return;
+
+  for (const model::Function &Function : Binary->Functions) {
+    auto &Summary = Oracle.at(Function.Entry);
+
+    abi::RegisterState::Map StateMap(Binary->Architecture);
+    initializeMapForDeductions(Summary, StateMap);
+
+    bool EnforceABIConformance = ABIEnforcement == FullABIEnforcement;
+    auto ResultMap = abi::applyRegisterStateDeductions(StateMap,
+                                                       Binary->DefaultABI,
+                                                       EnforceABIConformance);
+
+    if (!ResultMap.has_value())
+      continue;
+
+    for (const auto &[Register, State] : *ResultMap) {
+      llvm::StringRef Name = model::Register::getCSVName(Register);
+      if (llvm::GlobalVariable *CSV = M.getGlobalVariable(Name, true)) {
+        auto MaybeArg = State.IsUsedForPassingArguments;
+        auto MaybeRV = State.IsUsedForReturningValues;
+
+        if (MaybeArg == State::Yes || MaybeArg == State::YesOrDead)
+          if (Summary.ABIResults.ArgumentsRegisters.count(CSV) != 0)
+            Summary.ABIResults.ArgumentsRegisters[CSV] = State::Yes;
+
+        if (MaybeArg == State::No || MaybeArg == State::NoOrDead)
+          if (Summary.ABIResults.ArgumentsRegisters.count(CSV) != 0)
+            Summary.ABIResults.ArgumentsRegisters[CSV] = State::No;
+
+        if (MaybeRV == State::Yes || MaybeRV == State::YesOrDead)
+          if (Summary.ABIResults.FinalReturnValuesRegisters.count(CSV) != 0)
+            Summary.ABIResults.FinalReturnValuesRegisters[CSV] = State::Yes;
+
+        if (MaybeRV == State::No || MaybeRV == State::NoOrDead)
+          if (Summary.ABIResults.FinalReturnValuesRegisters.count(CSV) != 0)
+            Summary.ABIResults.FinalReturnValuesRegisters[CSV] = State::No;
```

Comment:
I feel like we're discarding any information about `Dead` registers. Is it done on purpose, if so why?

Reply:
Yeah, a hundred percent, it was still under experiment.